File: scripts/validate.ts
/**
 * Phase 1 자체 검증 스크립트.
 *  1) 데이터 무결성: id 중복, 날짜 포맷, 카테고리 enum 검증
 *  2) 각 난이도별 출제 분포 시뮬레이션 (1,000회)
 *  3) 한 세션에서 연승 20문제까지 사건 중복 없이 뽑히는지 확인
 *  4) 정답 검증 헬퍼 동작 점검
 */

import {
  DIFFICULTY_RANGES,
  createQuestionGenerator,
  generateQuestion,
  getAllEvents,
  isCorrectAnswer,
} from "../lib/gameLogic";
import type { Category, Difficulty, GameEvent } from "../lib/types";

const VALID_CATEGORIES: Category[] = [
  "history",
  "economy",
  "culture",
  "sports",
  "politics",
  "disaster",
  "science",
];

function assert(cond: boolean, msg: string) {
  if (!cond) {
    console.error("✗ FAIL:", msg);
    process.exitCode = 1;
  }
}

function checkDataIntegrity(events: GameEvent[]) {
  console.log(`\n[1] 데이터 무결성 검사 — 총 ${events.length}개 사건`);
  const ids = new Set<number>();
  const isoDate = /^\d{4}-\d{2}-\d{2}$/;
  let categoryCount: Record<string, number> = {};

  for (const e of events) {
    assert(!ids.has(e.id), `중복 id: ${e.id} (${e.title})`);
    ids.add(e.id);
    assert(isoDate.test(e.date), `날짜 포맷 오류: ${e.title} → ${e.date}`);
    assert(
      VALID_CATEGORIES.includes(e.category),
      `잘못된 카테고리: ${e.title} → ${e.category}`
    );
    assert(e.title.length > 0, `빈 제목: id=${e.id}`);
    categoryCount[e.category] = (categoryCount[e.category] ?? 0) + 1;
  }

  console.log("  카테고리 분포:", categoryCount);
  const years = events.map((e) => parseInt(e.date.slice(0, 4), 10));
  console.log(`  연도 범위: ${Math.min(...years)} ~ ${Math.max(...years)}`);
}

function simulateDistribution(difficulty: Difficulty, runs = 1000) {
  const range = DIFFICULTY_RANGES[difficulty];
  let success = 0;
  let outOfRange = 0;
  let wrongAnswerLogic = 0;
  const yearDiffs: number[] = [];

  for (let i = 0; i < runs; i++) {
    const q = generateQuestion(difficulty);
    if (!q) continue;
    success++;
    yearDiffs.push(q.yearDiff);
    if (q.yearDiff < range.minYearDiff || q.yearDiff >= range.maxYearDiff) {
      outOfRange++;
    }
    // earlierEventId가 실제로 더 이른 날짜인지 확인
    const earlier = [q.eventA, q.eventB].find((e) => e.id === q.earlierEventId)!;
    const other = q.eventA.id === earlier.id ? q.eventB : q.eventA;
    if (Date.parse(earlier.date) >= Date.parse(other.date)) {
      wrongAnswerLogic++;
    }
  }

  const avg =
    yearDiffs.length > 0
      ? (yearDiffs.reduce((a, b) => a + b, 0) / yearDiffs.length).toFixed(2)
      : "n/a";
  console.log(
    `  ${difficulty.padEnd(6)} | ${success}/${runs} 성공 | 평균 yearDiff=${avg} | 범위 이탈=${outOfRange} | 정답 로직 오류=${wrongAnswerLogic}`
  );
  assert(success === runs, `${difficulty}: 일부 출제 실패`);
  assert(outOfRange === 0, `${difficulty}: yearDiff 범위 이탈 발생`);
  assert(wrongAnswerLogic === 0, `${difficulty}: earlierEventId 산정 오류`);
}

function simulateSession() {
  console.log(`\n[3] 한 세션에서 연속 출제 (사건 중복 방지)`);
  for (const diff of ["easy", "normal", "hard"] as Difficulty[]) {
    const gen = createQuestionGenerator(diff);
    const seen = new Set<number>();
    let rounds = 0;
    for (let i = 0; i < 30; i++) {
      const q = gen.next();
      if (!q) break;
      assert(!seen.has(q.eventA.id), `${diff}: A 중복 ${q.eventA.id}`);
      assert(!seen.has(q.eventB.id), `${diff}: B 중복 ${q.eventB.id}`);
      seen.add(q.eventA.id);
      seen.add(q.eventB.id);
      rounds++;
    }
    console.log(`  ${diff.padEnd(6)} | 중복 없이 ${rounds}라운드 진행`);
    assert(rounds >= 5, `${diff}: 최소 연승 5라운드도 못 채움 (데이터 부족)`);
  }
}

function checkAnswerHelper() {
  console.log(`\n[4] 정답 판정 헬퍼 동작 확인`);
  const q = generateQuestion("easy");
  if (!q) {
    assert(false, "easy 난이도 출제 실패");
    return;
  }
  const wrongId = q.eventA.id === q.earlierEventId ? q.eventB.id : q.eventA.id;
  assert(isCorrectAnswer(q, q.earlierEventId), "정답 판정 실패");
  assert(!isCorrectAnswer(q, wrongId), "오답인데 정답으로 판정");
  console.log(`  ✓ isCorrectAnswer 정상 동작`);
}

function main() {
  const events = getAllEvents();
  checkDataIntegrity(events);

  console.log(`\n[2] 난이도별 출제 분포 (각 1,000회)`);
  simulateDistribution("easy");
  simulateDistribution("normal");
  simulateDistribution("hard");

  simulateSession();
  checkAnswerHelper();

  console.log(
    process.exitCode === 1
      ? "\n❌ 검증 실패 — 위 로그 확인 필요"
      : "\n✅ 모든 검증 통과"
  );
}

main();

File: scripts/validate.py
"""
Phase 1 검증 — Node가 설치되지 않은 환경에서도 데이터/알고리즘이 잘 도는지 확인하기 위한
TypeScript 미러 검증 스크립트. lib/gameLogic.ts 와 동일한 출제 알고리즘을 Python으로 재현한다.
실제 런타임 동작은 Phase 2 (Next.js) 단계에서 자동으로 검증됨.
"""

from __future__ import annotations
import json
import random
import re
from collections import Counter
from datetime import datetime
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
EVENTS = json.loads((ROOT / "data" / "events.json").read_text(encoding="utf-8"))["events"]

VALID_CATEGORIES = {"history", "economy", "culture", "sports", "politics", "disaster", "science"}
DIFFICULTY_RANGES = {
    "easy":   (10, 100),
    "normal": (5, 10),
    "hard":   (0, 5),
}

failures: list[str] = []
def expect(cond: bool, msg: str) -> None:
    if not cond:
        failures.append(msg)
        print("  ✗", msg)


def year(e: dict) -> int:
    return int(e["date"][:4])


def ts(e: dict) -> int:
    return int(datetime.strptime(e["date"], "%Y-%m-%d").timestamp())


# ─── 1. 데이터 무결성 ────────────────────────────────────────────
print(f"[1] 데이터 무결성 — 총 {len(EVENTS)}개 사건")
ids = set()
date_re = re.compile(r"^\d{4}-\d{2}-\d{2}$")
for e in EVENTS:
    expect(e["id"] not in ids, f"중복 id: {e['id']}")
    ids.add(e["id"])
    expect(bool(date_re.match(e["date"])), f"날짜 포맷 오류: {e['title']} → {e['date']}")
    expect(e["category"] in VALID_CATEGORIES, f"카테고리 오류: {e['title']} → {e['category']}")
    expect(len(e["title"]) > 0, f"빈 제목: id={e['id']}")
print("  카테고리 분포:", dict(Counter(e["category"] for e in EVENTS)))
years = [year(e) for e in EVENTS]
print(f"  연도 범위: {min(years)} ~ {max(years)}")


# ─── 2. 난이도별 출제 분포 ──────────────────────────────────────
def generate_question(difficulty: str, exclude: set[int] | None = None):
    exclude = exclude or set()
    pool = [e for e in EVENTS if e["id"] not in exclude]
    if len(pool) < 2:
        return None
    lo, hi = DIFFICULTY_RANGES[difficulty]
    order = pool[:]
    random.shuffle(order)
    for base in order:
        by = year(base)
        bts = ts(base)
        cands = [
            o for o in pool
            if o["id"] != base["id"]
            and lo <= abs(year(o) - by) < hi
            and ts(o) != bts
        ]
        if not cands:
            continue
        partner = random.choice(cands)
        earlier = base if bts < ts(partner) else partner
        a, b = (base, partner) if random.random() < 0.5 else (partner, base)
        return {
            "eventA": a, "eventB": b,
            "earlierEventId": earlier["id"],
            "yearDiff": abs(year(partner) - by),
        }
    return None


print("\n[2] 난이도별 출제 분포 (각 1,000회)")
for diff in ["easy", "normal", "hard"]:
    lo, hi = DIFFICULTY_RANGES[diff]
    success = out_of_range = wrong_logic = 0
    diffs: list[int] = []
    for _ in range(1000):
        q = generate_question(diff)
        if not q:
            continue
        success += 1
        diffs.append(q["yearDiff"])
        if not (lo <= q["yearDiff"] < hi):
            out_of_range += 1
        earlier = next(e for e in (q["eventA"], q["eventB"]) if e["id"] == q["earlierEventId"])
        other = q["eventB"] if earlier["id"] == q["eventA"]["id"] else q["eventA"]
        if ts(earlier) >= ts(other):
            wrong_logic += 1
    avg = sum(diffs) / len(diffs) if diffs else 0
    print(f"  {diff:6s} | {success}/1000 성공 | 평균 yearDiff={avg:.2f} | 범위 이탈={out_of_range} | 정답 로직 오류={wrong_logic}")
    expect(success == 1000, f"{diff}: 일부 출제 실패 ({success}/1000)")
    expect(out_of_range == 0, f"{diff}: yearDiff 범위 이탈 {out_of_range}건")
    expect(wrong_logic == 0, f"{diff}: earlierEventId 산정 오류 {wrong_logic}건")


# ─── 3. 한 세션에서 연속 출제 시 사건 중복 방지 ─────────────────
print("\n[3] 한 세션에서 연속 출제 (중복 방지)")
for diff in ["easy", "normal", "hard"]:
    used: set[int] = set()
    rounds = 0
    for _ in range(30):
        q = generate_question(diff, exclude=used)
        if not q:
            break
        expect(q["eventA"]["id"] not in used and q["eventB"]["id"] not in used,
               f"{diff}: 사건 중복 발생")
        used.add(q["eventA"]["id"])
        used.add(q["eventB"]["id"])
        rounds += 1
    print(f"  {diff:6s} | 중복 없이 {rounds}라운드 진행")
    expect(rounds >= 5, f"{diff}: 최소 연승 5라운드도 못 채움")


# ─── 결과 ────────────────────────────────────────────────────────
print()
if failures:
    print(f"❌ 검증 실패 — {len(failures)}건")
    for f in failures:
        print("  -", f)
    raise SystemExit(1)
else:
    print("✅ 모든 검증 통과")
